test/integration/scheduler_perf/scheduler_perf_test.go

```diff
@@ -51,258 +54,518 @@ var (
-		b.Run(name, func(b *testing.B) {
-			for feature, flag := range test.FeatureGates {
-				defer featuregatetesting.SetFeatureGateDuringTest(b, utilfeature.DefaultFeatureGate, feature, flag)()
+	dataItems := DataItems{Version: "v1"}
+	for _, tc := range testCases {
+		b.Run(tc.Name, func(b *testing.B) {
+			for _, w := range tc.Workloads {
+				b.Run(w.Name, func(b *testing.B) {
+					for feature, flag := range tc.FeatureGates {
+						defer featuregatetesting.SetFeatureGateDuringTest(b, utilfeature.DefaultFeatureGate, feature, flag)()
+					}
+					dataItems.DataItems = append(dataItems.DataItems, runWorkload(b, tc, w)...)
+				})
 			}
-			dataItems.DataItems = append(dataItems.DataItems, perfScheduling(test, b)...)
 		})
 	}
 	if err := dataItems2JSONFile(dataItems, b.Name()); err != nil {
 		klog.Fatalf("%v: unable to write measured data: %v", b.Name(), err)
 	}
 }
 
-func perfScheduling(test testCase, b *testing.B) []DataItem {
+func runWorkload(b *testing.B, tc *testCase, w *workload) []DataItem {
+	var mu sync.Mutex
+	var dataItems []DataItem
 	finalFunc, podInformer, clientset := mustSetupScheduler()
 	defer finalFunc()
 
-	nodePreparer, err := getNodePreparer(test.Nodes, clientset)
-	if err != nil {
-		b.Fatal(err)
-	}
-	if err := nodePreparer.PrepareNodes(); err != nil {
-		b.Fatal(err)
-	}
-	defer nodePreparer.CleanupNodes()
+	numPodsScheduledPerNamespace := make(map[string]int)
+	numNodes := 0
 
-	total := 0
-	for _, p := range test.InitPods {
-		if err := createPods(setupNamespace, p, clientset); err != nil {
-			b.Fatal(err)
-		}
-		total += p.Num
-	}
-	if !test.SkipWaitUntilInitPodsScheduled {
-		if err := waitNumPodsScheduled(b, total, podInformer, setupNamespace); err != nil {
+	b.ResetTimer()
+	for opIndex, op := range tc.WorkloadTemplate {
+		realOp, err := op.realOp.patchParams(w)
+		if err != nil {
 			b.Fatal(err)
 		}
-	}
-
-	// start benchmark
-	b.ResetTimer()
-
-	// Start test data collectors.
-	stopCh := make(chan struct{})
-	collectors := getTestDataCollectors(test, podInformer, b)
-	for _, collector := range collectors {
-		go collector.run(stopCh)
-	}
+		switch concreteOp := realOp.(type) {
+		case *createNodesOp:
+			nodePreparer, err := getNodePreparer(fmt.Sprintf("node-%d-", opIndex), concreteOp, clientset)
+			if err != nil {
+				b.Fatal(err)
+			}
+			if err := nodePreparer.PrepareNodes(numNodes); err != nil {
+				b.Fatal(err)
+			}
+			if numNodes == 0 {
+				// Schedule a cleanup at most once. The CleanupNodes function will list
+				// and delete *all* nodes.
+				// TODO(#93794): make CleanupNodes only clean up its own nodes to make
+				// this more intuitive?
+				defer nodePreparer.CleanupNodes()
+			}
+			numNodes += concreteOp.CreateNodes.(int)
+
+		case *createPodsOp:
+			var namespace string
+			if concreteOp.Namespace != nil {
+				namespace = *concreteOp.Namespace
+			} else {
+				namespace = fmt.Sprintf("namespace-%d", opIndex)
+			}
+			var stopCh chan struct{}
```

Comment:
Use `context.WithCancel` instead.
Use a parent context for the whole test and call it's cancel function from `b.Cleanup`

Reply:
Done.